crates/network/src/ethereum/builder.rs

```diff
@@ -126,48 +127,128 @@ where
     N: Network,
     N::UnsignedTx: From<TxLegacy> + From<TxEip1559> + From<TxEip2930> + From<TxEip4844Variant>,
 {
-    match (
-        request.access_list.as_ref(),
-        request.max_fee_per_blob_gas.as_ref(),
-        request.blob_versioned_hashes.as_ref(),
-        request.sidecar.as_ref(),
-    ) {
-        (_, Some(_), None, _) => Err(TransactionBuilderError::MissingKey("blob_versioned_hashes")),
-        (_, Some(_), _, None) => Err(TransactionBuilderError::MissingKey("sidecar")),
-        (Some(_), _, _, Some(_)) => {
-            Err(TransactionBuilderError::SidecarAndAccessListMutuallyExclusive)
+    if will_build_4844(&request)? {
+        build_4844(request).map(TxEip4844Variant::from).map(Into::into)
+    } else if will_build_2930(&request)? {
+        build_2930(request).map(Into::into)
+    } else if will_build_legacy(&request)? {
+        build_legacy(request).map(Into::into)
+    } else {
+        build_1559(request).map(Into::into)
+    }
+}
+
+fn get_invalid_common_fields(request: &TransactionRequest) -> InvalidTransactionRequestErrors {
+    let mut errors = vec![];
+
+    if request.nonce.is_none() {
+        errors.push(InvalidTransactionRequestError::MissingKey("nonce"));
+    }
+
+    if request.gas.is_none() {
+        errors.push(InvalidTransactionRequestError::MissingKey("gas_limit"));
+    }
+
+    InvalidTransactionRequestErrors(errors)
+}
+
+fn get_invalid_1559_fields(request: &TransactionRequest) -> Vec<InvalidTransactionRequestError> {
+    let mut errors = vec![];
+
+    if request.max_priority_fee_per_gas.is_none() {
+        errors.push(InvalidTransactionRequestError::MissingKey("max_priority_fee_per_gas"));
+    }
+
+    if request.max_fee_per_gas.is_none() {
+        errors.push(InvalidTransactionRequestError::MissingKey("max_fee_per_gas"));
+    }
+
+    errors
+}
+
+fn will_build_legacy(request: &TransactionRequest) -> Result<bool, TransactionBuilderError> {
+    if request.gas_price.is_some() {
```

Comment:
would prefer inverting and early-return these `if`s to avoid nesting

```
if request.gas_price.is_none { 
    return Ok(false0
}

let errors = get_invalid_common_fields(request);
if errors.is_empty() {
    ....
} else {
    ....
}

```